File: vt5/js/vuokraamo.js
//vuokraamo.js

"use strict"; // estää pahimpia virheitä

/* ajax_latauslista on array, johon kerätään kaikkien aliohjelmien 
* tunniste, jotka ovat lähettäneet ajax-kutsun, kutsun päätyttyä 
* kyseenomaisen ohjelman tunniste poistetaan listalta.
*/
var ajax_latauslista = [];

/*	window.onload = function()
* -------------------------------------------------------------------
*	Elokuvavuokraamo
*/
window.onload = function() {
	onko_ok();
	hae_vuokraukset();
	hae_vuokraajat();
	hae_elokuvat();
	hae_lajityypit();
	
	$( '#vuokrausPVM' ).on("change", tarkistaVPVM);
	$( '#palautusPVM' ).on("change", tarkistaPPVM);
	$( '#summa' ).on("change", tarkistaSumma);
	$( '#laheta' ).on("click", laheta_tiedot);
	$( '#kirjaudu' ).on("click", kirjaudu);
	$( '#lisaa_e' ).on("click", lisaa_elokuva);
	$( '#poista_e' ).on("click", poista_elokuva);
	olenko_kirjautunut();
};

/* hae_lajityypit()
* -------------------------------------------------------------------
*    Haetaan ajaxilla python-skriptiltä tiedot lajityypeistä.
*/
function hae_lajityypit() {
	ajax_latauslista.push('lajityyppi_haku');
	$( '#Loader' ).removeClass("hidden");
	$.ajax({
		async: true,
		url: "/~aakaneis/cgi-bin/vt5/flask.cgi/hae_lajityypit",
		type: "GET",
		dataType: "json",
		success: lisaa_lajityypit,
		error: ajax_virhe
	});
};
/* poista_latauslistalta(hakusana)
* -------------------------------------------------------------------
*    "Poistetaan" ajaxin latauslistalta hakusana, eli korvataan se tyhjällä
*    merkkijonolla, sekä tarkistetaan, onko latauslistalla ladattavaa.
*    Jos on, niin näytetään latausikoni, muuten piilotetaan se.
*/
function poista_latauslistalta(hakusana) {
	var ladattavia = false;
	var poistettu = false;
	for (var i = 0; i<ajax_latauslista.length; i++)
	{
		if (ajax_latauslista[i] == hakusana && !poistettu)
		{
			poistettu = true;
			ajax_latauslista[i] = '';
		}
		if (ajax_latauslista[i] != '')
			ladattavia = true;
	}
	if (ladattavia)
		$( '#Loader' ).removeClass("hidden");
	else
		$( '#Loader' ).attr("class","hidden");
}

/* lisaa_lajityypit(data, textStatus, request)
* -------------------------------------------------------------------
*    Lisätään ajaxilla saadut lajityyppitiedot verkkosivun lomakkeen
*    select-elementtiin
*/
function lisaa_lajityypit(data, textStatus, request) {
	poista_latauslistalta('lajityyppi_haku');
	var lajityyppi = $( '#lajityyppi' );
	for (var i=0; i<data.length; i++) {
		lajityyppi.append( $("<option value=\"" + data[i].lid + "\">" + data[i].Nimi + "</option>") );
	}
};

/* olenko_kirjautunut()
* -------------------------------------------------------------------
*    Tarkistetaan Ajaxilla, onko käyttäjä kirjautunut. 
*/
function olenko_kirjautunut() {
	$( '#Loader' ).removeClass("hidden");
	ajax_latauslista.push('olenko_kirjautunut');
	$.ajax({
		async: true,
		url: "/~aakaneis/cgi-bin/vt5/flask.cgi/auth",
		type: 'GET',
		dataType: 'json',
		success: kirjautumisen_varmistus,
		error: ajax_virhe
	});
};

/* kirjaudu()
* -------------------------------------------------------------------
*    Toimitetaan ajaxilla pythonille kirjautumistiedot tarkistettavaksi.
*/
function kirjaudu(e) {
	ajax_latauslista.push('olenko_kirjautunut');
	$( '#Loader' ).removeClass("hidden");
	e.preventDefault();
	var postForm = {
			'ssana' : $( '#ssana' ).val(),
			'ktun' 	: $( '#ktun' ).val()
		};
		
		$.ajax({
			async: true,
			type	: 'POST',
			url		: "/~aakaneis/cgi-bin/vt5/flask.cgi/kirjaudu",
			data	: postForm,
			dataType: 'json',
			success: kirjautumisen_varmistus,
			error: ajax_virhe
		});
};

/* kirjautumisen_varmistus(data, textStatus, request)
* -------------------------------------------------------------------
*    Varmistetaan pythonin palauttamasta datasta että ollaan kirjauduttu
*    oikein, muuten näytetään virheet ja ei päästetä jatkamaan.
*/
function kirjautumisen_varmistus(data, textStatus, request) {
	poista_latauslistalta('olenko_kirjautunut');
	if (data.length == 0) {
		$( 'header' ).attr("class","hidden");
		$( '#Main' ).removeClass("hidden");
		$( '#Kirjaudu' ).attr("class","hidden");
	}
	else {
		$( 'header' ).removeClass("hidden");
		var virhe_UL = $( "header > ul" );
		virhe_UL.empty();
		for (var i=0; i<data.length; i++) {
			var virhe = $('<li>');
			virhe.append(data[i]);
			virhe_UL.append(virhe);
		}
	}
};

/* hae_vuokraajat()
* -------------------------------------------------------------------
*    Haetaan ajaxilla vuokraajat
*/
function hae_vuokraajat() {
	ajax_latauslista.push('vuokraaja_haku');
	$( '#Loader' ).removeClass("hidden");
	$.ajax({
		async: true,
		url: "/~aakaneis/cgi-bin/vt5/flask.cgi/hae_vuokraajat",
		type: "GET",
		dataType: "json",
		success: lisaa_vuokraajat,
		error: ajax_virhe
	});
};

/* lisaa_vuokraajat(data, textStatus, request)
* -------------------------------------------------------------------
*    Lisätään ajaxilla saadut vuokraajien tiedot verkkosivun lomakkeen
*    select -elementtiin
*/
function lisaa_vuokraajat(data, textStatus, request) {
	poista_latauslistalta('vuokraaja_haku');
	var vuokraaja = $( '#vuokraaja' );
	for (var i=0; i<data.length; i++) {
		vuokraaja.append( $("<option value=\"" + data[i].jid + "\">" + data[i].Nimi + "</option>") );
	}
};

/* hae_elokuvat()
* -------------------------------------------------------------------
*    Haetaan ajaxilla elokuvista tiedot
*/
function hae_elokuvat() {
	ajax_latauslista.push('elokuva_haku');
	$( '#Loader' ).removeClass("hidden");
	$.ajax({
		async: true,
		url: "/~aakaneis/cgi-bin/vt5/flask.cgi/hae_elokuvat",
		type: "GET",
		dataType: "json",
		success: lisaa_elokuvat,
		error: ajax_virhe
	});
};

/* lisaa_elokuvat(data, textStatus, request)
* -------------------------------------------------------------------
*    Lisätään ajaxilla saadut elokuvatiedot verkkosivun lomakkeen
*    select-elementtiin
*/
function lisaa_elokuvat(data, textStatus, request) {
	poista_latauslistalta('elokuva_haku');
	var elokuva = $( '#elokuva' );
	var elokuva_poistettava = $( '#elokuva_poista' );
	elokuva.empty();
	elokuva_poistettava.empty();
	for (var i=0; i<data.length; i++) {
		elokuva.append( $("<option value=\"" + data[i].eid + "\">" + data[i].Nimi + "</option>") );
		elokuva_poistettava.append( $("<option value=\"" + data[i].eid + "\">" + data[i].Nimi + "</option>") );
	}
};

/* hae_vuokraukset()
* -------------------------------------------------------------------
*    Hakee ajaxina vuokraukset tietokannasta.
*/
function hae_vuokraukset() {
	ajax_latauslista.push('vuokraus_haku');
	$( '#Loader' ).removeClass("hidden");
	$.ajax({
		async: true,
		url: "/~aakaneis/cgi-bin/vt5/flask.cgi/hae_vuokraukset",
		type: "GET",
		dataType: "json",
		success: lisaa_vuokraukset,
		error: ajax_virhe
	});
};

/* lisaa_vuokraukset(data, textStatus, request)
* -------------------------------------------------------------------
*    Muodostetaan ajaxilla saadusta datasta kaksitasoinen lista 
*    vuokrauksista.
*/
function lisaa_vuokraukset(data, textStatus, request) {
	poista_latauslistalta('vuokraus_haku');
	var main_ul = $( '#vuokrauslistaus' );
	main_ul.empty();
	var tiedot = data;
	for ( var i=0; i<tiedot.length; i++) {
		var vuokraajan_tiedot = tiedot[i];
		var vuokraaja_id = vuokraajan_tiedot.jid;
		var vuokraaja = vuokraajan_tiedot.Nimi;
		var lista_elementti = $('<li>');
		lista_elementti.append(vuokraaja);
		var vuokralista = $('<ul>');
		var vuokraukset = vuokraajan_tiedot.vuokraukset;
		for (var j=0;j<vuokraukset.length; j++) {
			var vuokraus = $('<li>');
			vuokraus.append( $("<a href=\"\">" + vuokraukset[j].elokuva + "</a>") );
			vuokraus.find( "a" ).on("click", { jid: vuokraaja_id, eid: vuokraukset[j].eid, vpvm: vuokraukset[j].vpvm, ppvm: vuokraukset[j].ppvm, summa: vuokraukset[j].summa } ,valmistele_muokkaus);
			vuokraus.append($('<label>').append(vuokraukset[j].vpvm));
			vuokraus.append($('<label>').append(vuokraukset[j].ppvm));
			vuokraus.append($('<label>').append(vuokraukset[j].summa));
			vuokralista.append(vuokraus);
		}
		lista_elementti.append(vuokralista);
		main_ul.append( lista_elementti );
	}
};

/* valmistele_muokkaus(e)
* -------------------------------------------------------------------
*    Muuttaa hieman vuokrauksenlisäys lomaketta, jotta se toimisi
*    vuokrauksen muokkaukseen.
*/
function valmistele_muokkaus(e) {
	e.preventDefault();
	$( '#laheta' ).attr("value","Muokkaa");
	$( '#muokattavan_eid' ).val( e.data.eid ).change();
	$( '#muokattavan_jid' ).val(e.data.jid ).change();
	$( '#muokattavan_vpvm' ).val(e.data.vpvm ).change();
	$( '#summa' ).val(e.data.summa ).change();
	$( '#vuokrausPVM' ).val(e.data.vpvm ).change();
	$( '#palautusPVM' ).val(e.data.ppvm ).change();
	$( '#vuokraaja' ).val(e.data.jid).change();
	$( '#elokuva' ).val(e.data.eid).change();
};

/* ajax_virhe(xhr, status, error)
* -------------------------------------------------------------------
*    Lokitetaan consoliin ajaxin antama virhe. 
*/
function ajax_virhe(xhr, status, error) {
	console.log( "Error: " + error );
	console.log( "Status: " + status );
	console.log( xhr );
	
	$( '#Loader' ).attr("class","hidden");
};

/* tarkistaPVM()
* -------------------------------------------------------------------
*   Tarkistaa onko vuokrauspäivämäärä annettu varmasti oikein.
*/
function tarkistaPPVM() {
	var apuid = $(this).attr("id");
	if ($(this).val() == "" || Date.parse($(this).val())) {
		$( '#' + apuid + "_req" ).attr("class", "hidden");
	}
	else {
		$( '#' + apuid + "_req" ).removeClass("hidden");
	}
	onko_ok();
};

/* poista_elokuva(e)
* -------------------------------------------------------------------
*    Ajaxilla kertoo pythonille, mikä elokuva tulisi poistaa listalta.
*/
function poista_elokuva(e) {
	ajax_latauslista.push('tiedon_lahetys');
	$( '#Loader' ).removeClass("hidden");
	e.preventDefault();
	var postForm = {
		'eid': $( '#elokuva_poista' ).val()
	};
	
	$.ajax({
		async: true,
		type: 'POST',
		url	: "/~aakaneis/cgi-bin/vt5/flask.cgi/poista_elokuva",
		data	: postForm,
		dataType: 'json',
		success: tiedot_lahetetty,
		error: ajax_virhe
	});
};

/* lisaa_elokuva(e)
* -------------------------------------------------------------------
*    Ajaxilla kertoo pythonille, millainen elokuva lisätään tietokantaan.
*/
function lisaa_elokuva(e) {
	ajax_latauslista.push('tiedon_lahetys');
	$( '#Loader' ).removeClass("hidden");
	e.preventDefault();
	var postForm = {
		'nimi'	: $( '#e_nimi' ).val(),
		'julkaisuvuosi' : $( '#julkaisuvuosi' ).val(),
		'hinta' : $( '#hinta' ).val(),
		'arvio' : $( '#arvio' ).val(),
		'lajityyppi' : $( '#lajityyppi' ).val()
	}
	$.ajax({
		async: true,
		type: 'POST',
		url	: "/~aakaneis/cgi-bin/vt5/flask.cgi/lisaa_elokuva",
		data	: postForm,
		dataType: 'json',
		success: tiedot_lahetetty,
		error: ajax_virhe
	});
};

/* laheta_tiedot(e) 
* -------------------------------------------------------------------
*    Lähetetään lomakkeen tiedot python skriptiin
*/
function laheta_tiedot(e) {
	ajax_latauslista.push('tiedon_lahetys');
	$( '#Loader' ).removeClass("hidden");
	e.preventDefault();
	if (onko_ok()) {
		var postForm = {
			'jid' 	: $( '#vuokraaja' ).val(),
			'eid' 	: $( '#elokuva' ).val(),
			'vpvm' 	: $( '#vuokrausPVM' ).val(),
			'ppvm' 	: $( '#palautusPVM' ).val(),
			'summa' : $( '#summa' ).val(),
			'jid_old'	: $( '#muokattavan_jid' ).val(),
			'eid_old'	: $( '#muokattavan_eid' ).val(),
			'vpvm_old'	: $( '#muokattavan_vpvm' ).val()
		};
		
		$.ajax({
			async: true,
			type	: 'POST',
			url		: "/~aakaneis/cgi-bin/vt5/flask.cgi/lisaa_vuokraaja",
			data	: postForm,
			dataType: 'json',
			success: tiedot_lahetetty,
			error: ajax_virhe
		});
		
	}
};

/* tiedot_lahetetty(data, textStatus, request)
* -------------------------------------------------------------------
*    Käsitellään palautunut data. data[0].success kertoo, onko 
*    tiedot käsitelty onnistuneesti, data[0].virheet sisältää taulun 
*    ilmenneistä ongelmista. 
*/
function tiedot_lahetetty(data, textStatus, request) {
	poista_latauslistalta('tiedon_lahetys');
	$( '#Loader' ).removeClass("hidden");
	if (data[0].success)
	{
		$( 'header' ).attr("class","hidden");
		$( '#muokattavan_eid' ).attr("value","");
		$( '#muokattavan_jid' ).attr("value","");
		$( '#muokattavan_vpvm' ).attr("value","");
		$( '#laheta' ).attr("value","Vuokraa!");
		hae_vuokraukset();
		hae_elokuvat();
	}
	else
	{
		$( 'header' ).removeClass("hidden");
		var virhe_UL = $( "header > ul" );
		virhe_UL.empty();
		for (var i=0; i<data[0].errors.length; i++) {
			var virhe = $('<li>');
			virhe.append(data[0].errors[i]);
			virhe_UL.append(virhe);
		}
	}
};

/* tarkistaPVM()
* -------------------------------------------------------------------
*   Tarkistaa onko vuokrauspäivämäärä annettu varmasti oikein.
*/
function tarkistaVPVM() {
	var apuid = $(this).attr("id");
	if (Date.parse($(this).val())) {
		$( '#' + apuid + "_req" ).attr("class", "hidden");
	}
	else {
		$( '#' + apuid + "_req" ).removeClass("hidden");
	}
	onko_ok();
};

/* tarkistaSumma()
* -------------------------------------------------------------------
*   Tarkistaa, onko annettu summa varmasti oikein.
*/
function tarkistaSumma() {
	var apuid = $(this).attr("id");
	var summa = parseInt($(this).val());
	if (!isNaN(summa) && summa >= 0 ) {
		$( '#' + apuid + "_req" ).attr("class", "hidden");
	}
	else {
		$( '#' + apuid + "_req" ).removeClass("hidden");
	}
	onko_ok();
}
/* onko_ok()
* -------------------------------------------------------------------
*   Katsoo, onko virheitä lomakkeessa ja tarvittaessa vaihtaa lähetä-
*   painikkeen pois käytöstä (ja vice versa)
*/
function onko_ok() {
	var span = $( "span" );
	for (var i=0; i<span.length; i++) {
		if ($(span[i]).attr("class") != "hidden") {
			$('#laheta').attr("disabled","disabled");
			return false;
		}
	}
	$('#laheta').removeAttr("disabled");
	return true;
}

File: vt5/vuokraamo.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from flask import Flask, jsonify, redirect, escape, request, Response, make_response, session
import sqlite3
import logging
import os
import sys
# korjaa tähän polku oikeaan paikkaan. lokia ei saa tehdä cgi-bin-kansion alaisuuteen
# varmista, että piilo-kansioon on kaikilla kirjoitusoikeus
# älä luo kansioita windowsin kautta vaan vain ja ainoastaan unix-komentoriviltä
logging.basicConfig(filename=os.path.abspath('../../piilo/vuokraamo.log'),level=logging.DEBUG)
app = Flask(__name__) 
app.secret_key = b'h6\x99\xd8Uq\x99i\x98\xcb\xf5\xf5-\x8d\xf1\xa0\xbc\xb8\xc6^|"\xf4\x98'
app.debug = True

# Koitetaan, löytyykö simplejsonia, jos ei niin lisätään json, jossa tietääkseni on oikeat 
# (tai siis ohjelman vaatimat) metodit.
try:
    import simplejson as json
except Exception as e:
    logging.debug("Simplejsonia ei löytynyt")
    logging.debug(str(e))
    try:
        import json
    except Exception as e:
        logging.debug("jsonia ei löytynyt")
        logging.debug(str(e))

# Kirjautumissivu
@app.route('/kirjaudu', methods=['POST', 'GET'])
def kirjaudu():
    # virheet -listaan kerätään ilmenneitä virheitä, jotka palautetaan sitten.
    # krypt_* tarkoittaa vain kryptattua muotoa tiedosta
    virheet = []
    krypt_salasana = b'\x18}P?+k[ g[\x90\x1f\xe5\x08\x9c%\xc6{\xff\x12\xd8\x9ec\xbf\x87\x8b\xdf\xc4\x89\x1e\xd4y\x04\xcf\x166\xda\xcb\x18\xcf]\xbc\x94\xd9\x97Ex\x11G& \x81\x18\xba\xdf\x99\x97\xfa\t\x0e2\x10\x8a&'
    krypt_tunnus = b'F\xf8Q\t\x90\xf5\xa3@\x96Dp\x91\xfd\x06\xc7\x8c9V\xf2!\xad\xf8\xf5\x96\xfdOD\xb2\xe0\xed\x1a\x97\xd5\xa1l\x12l\x95\xdd&3/\x80\xef0]\xce\xed\x945]"a\xd4\\y-\xba\x99N\x83#Z\xc4'
    # Kirjautuminen pähkinänkuoressa: suolataan s ja t, sha415:lla vedetään annettu
    # tunnus ja salasana ja verrataan sitten kryptattuihin arvoihin. 
    try:
        import hashlib
        s = hashlib.sha512()
        t = hashlib.sha512()
        s.update(app.secret_key)
        t.update(app.secret_key)
        try:
            salasana = request.form.get("ssana").encode('utf-8')
            tunnus = request.form.get("ktun").encode('utf-8')
        except:
            salasana = b""
            tunnus = b""
        s.update(salasana)
        t.update(tunnus)
        if t.digest() == krypt_tunnus and s.digest() == krypt_salasana:
            session['kirjautunut'] = "ok"
        else:
            virheet.append(u'Käyttäjätunnus ja/tai salasana oli väärin.')
    except Exception as e:
        logging.debug("Kirjautumishäikkä")
        logging.debug(str(e))
        virheet.append(u"Kirjautuminen oli hankalaa")
    resp = make_response(json.dumps(virheet), 200)
    resp.mimetype = "application/json"
    return resp

# kirjautumisen tarkistus
@app.route('/auth')
def auth():
    # Tämä on aika yksiselitteinen pätkä jo itsessään, onko sessionissa 'kirjautunut'
    # tietoa vai ei ja toimenpiteet sen mukaan.
    virheet = []
    try:
        if not 'kirjautunut' in session:
            virheet.append(u'Et ole kirjautunut sisään.')
    except:
        virheet.append(u'Et ole kirjautunut sisään.')
    resp = make_response(json.dumps(virheet), 200)
    resp.mimetype="application/json"
    return resp

# Hakee vuokraajat ja tulostaa ne JSON muodossa
@app.route('/hae_vuokraajat')
def hae_vuokraajat():
    try:
        vuokraajat = []
        con = yhdista_kantaan()
        cur = con.cursor()
        cur.execute("""SELECT JasenID, Nimi
        FROM Jasen
        """)
        # J tulee Jäsen -sanasta, olisi voinut käyttää myös v eli vuokraaja.
        # lisäillään vaan selectin tuottama tavara vuokraajat-listaan ja lähetetään eteenpäin.
        for j in cur:
            vuokraajat.append({"jid":j['JasenID'],"Nimi":j['Nimi']})
        resp = make_response(json.dumps(vuokraajat), 200)
        #resp.charset = "UTF-8"
        resp.mimetype = "application/json"
        con.close()
        return resp
    except Exception as e:
        logging.debug("Jokin meni hakiessa kummallisesti:")
        logging.debug(str(e))

# Hakee elokuvat ja tulostaa ne JSON muodossa
@app.route('/hae_elokuvat')
def hae_elokuvat():
    try:
        elokuvat = []
        con = yhdista_kantaan()
        cur = con.cursor()
        cur.execute("""SELECT ElokuvaID, Nimi
        FROM Elokuva
        """)
        for e in cur:
            elokuvat.append({"eid":e['ElokuvaID'],"Nimi":e['Nimi']})
        resp = make_response(json.dumps(elokuvat), 200)
        #resp.charset = "UTF-8"
        resp.mimetype = "application/json"
        return resp
        con.close()
    except Exception as e:
        logging.debug("Jokin meni hakiessa kummallisesti:")
        logging.debug(str(e))

# Hakee elokuvalistauksen
@app.route('/hae_vuokraukset')
def hae_vuokraukset():
    vuokraajat = []
    vuokraukset = []
    logging.debug("Löysin vuokrauksiin")
    try:
        con = yhdista_kantaan()
        cur = con.cursor()
        cur.execute("""SELECT JasenID, Nimi
        FROM Jasen
        ORDER BY Nimi
        """)
        henkilot = cur.fetchall()
        # j tulee Jäsen -sanasta, olisi voinut käyttää selkeämpääkin muuttujaa.
        # Tässä seuraa vain seuraavaa:
        # rullataan selectillä haetut jäsenet, haetaan heidän JasenID:llä vuokraukset,
        # lisätään vuokraajan vuokraukset listaan ja laitetaan vuokraajat listaan sitten
        # Nimi, JasenID ja vuokraukset.
        for j in henkilot:
            vuokraukset = []
            try:
               sql = ''' SELECT v.JasenID as jid, v.ElokuvaID as eid, v.VuokrausPVM as vpvm, e.Nimi as elokuva, v.PalautusPVM as ppvm, v.Maksettu as summa
               FROM Vuokraus v, Elokuva e
               WHERE e.ElokuvaID = v.ElokuvaID
               AND v.JasenID = :jid
               ORDER BY v.VuokrausPVM DESC
               '''
               cur.execute(sql, {"jid": j["JasenID"]})
            except:
               logging.debug("En saanut jäsenen tietoa")
               logging.debug(sys.excinfo()[0])
            for v in cur:
                vuokraukset.append({"eid":v['eid'],"vpvm":v['vpvm'],"elokuva":v['elokuva'],"ppvm":v['ppvm'],"summa":v['summa']})
            vuokraajat.append({"jid":j['JasenID'],"Nimi":j['Nimi'],"vuokraukset":vuokraukset})
        con.close()
    except Exception as e:
        logging.debug("Jokin meni hassusti")
        logging.debug(str(e))
    resp = make_response(json.dumps(vuokraajat), 200)
    #resp.charset = "UTF-8"
    resp.mimetype = "application/json"
    return resp
    
# Lajityyppien haku
@app.route('/hae_lajityypit')
def hae_lajityypit():
    try:
        lajityypit = []
        con = yhdista_kantaan()
        cur = con.cursor()
        cur.execute("""SELECT LajityyppiID, Tyypinnimi
        FROM Lajityyppi
        """)
        # l tarkoittaa lajityyppiä. Ei pitäisi tämän koodin olla kovin epäselvä.
        for l in cur:
            lajityypit.append({"lid":l['LajityyppiID'],"Nimi":l['Tyypinnimi']})
        resp = make_response(json.dumps(lajityypit), 200)
        #resp.charset = "UTF-8"
        resp.mimetype = "application/json"
        return resp
        con.close()
    except Exception as e:
        logging.debug("Jokin meni hakiessa kummallisesti:")
        logging.debug(str(e))
        
# Lisää vuokraaja
@app.route('/lisaa_vuokraaja', methods=['POST', 'GET'])
def lisaa_vuokraaja():
    # virhelistan alustus, tähän kerätään ilmenneet virheet.
    # Tarkistetaan heti aluksi jo, että ollaanko kirjauduttu, ettei tätä 
    # voisi käyttää ilman kirjautumista.
    virheet = []
    try:
        if not 'kirjautunut' in session:
            virheet.append(u'Et ole kirjautunut sisään.')
    except:
        virheet.append(u'Et ole kirjautunut sisään.')
    try:
        con = yhdista_kantaan()
        cur = con.cursor()
        try:
            # Tarkastetaan, onko muokkaukseen vaadittavia tietoja, jos ei ole niin silloin lisätään data.
            if request.form.get("jid_old")=='' or request.form.get("eid_old")=='' or request.form.get('vpvm_old')=='':
                sql= '''INSERT INTO Vuokraus (JasenID,ElokuvaID,VuokrausPVM,PalautusPVM,Maksettu)
                VALUES (:jid,:eid,:vpvm,:ppvm,:summa)
                '''
                # Ei executeta, ellei olla kirjautuneita
                if 'kirjautunut' in session:
                    cur.execute(sql, {"jid": request.form.get("jid"), "eid": request.form.get("eid"), "vpvm": request.form.get("vpvm"), "ppvm": request.form.get("ppvm"), "summa": request.form.get("summa")})
            # Muuten päivitetään olemassa olevaa.
            else:
                sql = '''UPDATE vuokraus
                SET JasenID = :jid,
                ElokuvaID = :eid,
                VuokrausPVM = :vpvm,
                PalautusPVM = :ppvm,
                Maksettu = :summa
                WHERE JasenID = :jid_old
                AND ElokuvaID = :eid_old
                AND VuokrausPVM = :vpvm_old
                '''
                # Taas ei ajeta käskyä ilman, että ollaan kirjautuneita.
                if 'kirjautunut' in session:
                    cur.execute(sql,  {"jid": request.form.get("jid"), "eid": request.form.get("eid"), "vpvm": request.form.get("vpvm"), "ppvm": request.form.get("ppvm"), "summa": request.form.get("summa"), "jid_old": request.form.get("jid_old"), "eid_old": request.form.get("eid_old"), "vpvm_old": request.form.get("vpvm_old")})
        except Exception as e:
            logging.debug("insert ei toiminut.")
            logging.debug(str(e))
            con.rollback()
            virheet.append(u"Annetuissa tiedoissa oli virhe. Esim. Vastaava syöte löytyy jo tietokannasta tai tiedot olivat virheellisiä (mm. Vuokrauspäivämäärä ennen palautuspäivämäärää)")
        con.commit()
        con.close()
    except Exception as e:
        logging.debug("En saanut yhteyttä kantaan.")
        logging.debug(str(e))
        virheet.append(u"Tietokanta ei auennut.")
    # Kerätään data-listaan onnistuiko ja laitetaan perään vielä ilmenneet virheet.
    data = []
    data.append({"success": len(virheet) == 0, "errors": virheet})
    resp = make_response(json.dumps(data), 200)
    resp.mimetype = "application/json"
    return resp

# Lisää elokuva
@app.route('/lisaa_elokuva', methods=['POST', 'GET'])
def lisaa_elokuva():
    # virheet -listan alustus
    virheet=[]
    
    # Tarkistetaan kirjautuminen
    try:
        if not 'kirjautunut' in session:
            virheet.append(u'Et ole kirjautunut sisään.')
    except:
        virheet.append(u'Et ole kirjautunut sisään.')
    try:
        con = yhdista_kantaan()
        cur = con.cursor()
        try: 
            # Luodaan uusi elokuva, jos tiedot on tulleet oikein
            sql= '''
            INSERT INTO Elokuva (Nimi, Julkaisuvuosi, Vuokrahinta, Arvio, LajityyppiID)
            VALUES (:nimi, :julkaisuvuosi, :hinta, :arvio, :lajityyppi)
            '''
            # Ajetaan execute vain, jos ollaan kirjautuneita
            if 'kirjautunut' in session:
                cur.execute(sql, {"nimi": request.form.get("nimi"), "julkaisuvuosi": request.form.get("julkaisuvuosi"), "hinta": request.form.get("hinta"), "arvio": request.form.get("arvio"), "lajityyppi": request.form.get("lajityyppi")})
        except Exception as e:
            logging.debug("En saanut elokuvaa lisättyä.")
            logging.debug(str(e))
            con.rollback()
            virheet.append(u"Elokuvaa ei kyetty lisäämään. Virheellinen syöte?")
        con.commit()
        con.close()
    except Exception as e:
        logging.debug("En saanut yhteyttä kantaan.")
        logging.debug(str(e))
        virheet.append(u'Tietokanta ei aukea.')
    # Sama palautus kuin edellisessä.
    data = []
    data.append({"success": len(virheet) == 0, "errors": virheet})
    resp = make_response(json.dumps(data), 200)
    resp.mimetype = "application/json"
    return resp

# Poista annettu elokuva
@app.route('/poista_elokuva', methods=['POST', 'GET'])
def poista_elokuva():
    virheet=[]
    
    try:
        con = yhdista_kantaan()
        cur = con.cursor()
        try:
            # Poistetaan eka kaikki vuokraukset, joissa kys. om. elokuva mainitaan
            sql = ''' DELETE FROM Vuokraus
            WHERE ElokuvaID = :eid
            '''
            if 'kirjautunut' in session:
                cur.execute(sql, {"eid": request.form.get("eid")})
            # Poistetaan sitten elokuva, kun tiedetään ettei muut voi viitata elokuvaID:hen
            sql = ''' DELETE FROM Elokuva
            WHERE ElokuvaID = :eid
            '''
            if 'kirjautunut' in session:
                cur.execute(sql, {"eid": request.form.get("eid")})
        except Exception as e:
            logging.debug("En saanut elokuvaa poistettua.")
            logging.debug(str(e))
            con.rollback()
            virheet.append(u"Elokuvaa ei kyetty poistamaan.")
        con.commit()
        con.close()
    except Exception as e:
        logging.debug("En saanut yhteyttä kantaan.")
        logging.debug(str(e))
        virheet.append(u'Tietokanta ei aukea')
    
    data = []
    data.append({"success": len(virheet) == 0, "errors": virheet})
    resp = make_response(json.dumps(data), 200)
    resp.mimetype = "application/json"
    return resp

# Yhdistää tietokantaan
def yhdista_kantaan():
    # sqlite haluaa absoluuttisen polun
    try:
        con = sqlite3.connect(os.path.abspath('../../piilo/video'))
        con.row_factory = sqlite3.Row
        return con;
    except Exception as e:
        logging.debug("jotain meni pieleen")
        logging.debug(str(e))


if __name__ == '__main__':
    app.debug = True
    app.run(debug=True)